### PR TITLE
Move certificate generation to hosts

### LINKED ISSUE #1435: TLS certificates should be valid for *.service and *.node
When working on #1406, I used the short DNS form for a service, i.e. without the `consul_dns_domain`, and got the following message:
```
May 10 07:33:01 lb0-control-01 docker[21812]: time="2016-05-10T07:33:01Z" level=error msg="Put https://consul.service:8500/v1/session/create: x509: certificate is valid for localhost, *.node.consul, *.service.consul, not consul.service"
```
The cert we deploy should also be valid for those short constructions. 

- Ansible version (`ansible --version`): 1.9.4
- Python version (`python --version`): n/a
- Git commit hash or branch: master
- Cloud Environment: aws
- Terraform version (`terraform version`): 0.6.14

### PR DESCRIPTION
**Pull request by siddharthist**
- [x] Installs cleanly on a fresh build of most recent master branch
- [x] Upgrades cleanly from the most recent release
- [x] Updates documentation relevant to the changes

TLS protection for Docker over TLS is now on by default.

This will pave the road for other TLS dependent components, such as Docker Swarm.

Acceptance tests: 
Make sure that the output of 
```
openssl x509 -in /etc/pki/mantl/cert -noout -text
```
looks right.
 
With Docker TCP enabled:
Without TLS:
```
$ sudo docker -H :2375 info
$ sudo docker -H :2375 run hello-world
```

With Docker TCP+TLS:
```
$ sudo docker -H :2375 --tlsverify --tlscacert=/etc/pki/CA/ca.cert --tlscert=/etc/pki/mantl/cert --tlskey=/etc/pki/mantl/key info
$ sudo docker -H :2375 --tlsverify --tlscacert=/etc/pki/CA/ca.cert --tlscert=/etc/pki/mantl/cert --tlskey=/etc/pki/mantl/key run hello-world
```

Fixes #1435, #1459

### CODE REVIEW

File: roles/consul/tasks/nginx_proxy.yml [hunk 1]
```diff
@@ -1,9 +1,20 @@
...
+  file:
+    state: directory
+    path: "{{ item }}"
+  with_items:
+    - /etc/nginx
+    - /etc/nginx/templates
+  tags:
+    - consul
+
 - name: deploy nginx template
```
Comment (ChrisAubuchon) [inline roles/consul/tasks/nginx_proxy.yml]: Should probably do this to `marathon/nginx_proxy.yml` as well to keep things consistent.

**Comment by siddharthist** [comment]
Wow a green travis build :star2:

File: roles/consul/tasks/nginx_proxy.yml (hunk 1 above)
Comment (siddharthist) [inline roles/consul/tasks/nginx_proxy.yml]: Done!

**Comment by siddharthist** [comment]
Quota failures on AWS

**Comment by sehqlr** [comment]
I got this to work, with two exceptions:
1. I had to `daemon-reload` and restart docker to load in the new config after switching `docker_tcp` to true
2. the TCP connection only takes encrypted traffic, but the acceptance test stated above expects unencrypted requests to work as well

If you look at the documentation for the [docker daemon socket option](https://docs.docker.com/v1.10/engine/reference/commandline/daemon/#daemon-socket-option), and compare it to [the docker role defaults file](https://github.com/CiscoCloud/mantl/blob/c71f83ac06c5ac9de7a74c35e1ae5ea9651958e9/roles/docker/defaults/main.yml), I think that this config logically results in point 2.

While this conflicts with the description of the accepted behavior, I think that it's better if we *don't* have an unencrypted connection open. Is there a reason why it is necessary?

**Comment by siddharthist** [comment]
@sehqlr I think we've miscommunicated again :/ the commands without TLS should only work when `docker_tcp_tls` is set to `false`, but they should work in that case.

I'll get to work on a daemon-reload handler! Thanks for pointing this out!

**Comment by sehqlr** [comment]
Anything to help out :)

**Comment by siddharthist** [comment]
@sehqlr Actually, that handler is totally present. The task that changes the docker network options notifies `reload docker` which is exactly a `daemon-reload` plus a `restart docker`, so I wonder what went wrong that you had to do that manually...

**Comment by sehqlr** [comment]
Yeah, I'm not sure either. It's entirely possible that I goofed it up during the upgrade process.

File: roles/certificates/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    state: present
+
+- name: create certificate destination directory
+  sudo: yes
+  file:
+    state: directory
+    path: "{{ destination }}"
+    owner: root
+    group: root
+    mode: 755
```
Comment (ChrisAubuchon) [inline roles/certificates/tasks/main.yml]: This needs to be `0755`. It's not getting parsed as octal and creates the directory:
```
d-wxrw--wt. 2 root root   41 Jun  7 20:51 mantl
```

File: roles/certificates/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
+---
+- name: deploy CA public and private keys
+  sudo: yes
+  copy:
+    src: "{{ item.src }}"
+    dest: "{{ item.dest }}"
+    owner: root
```
Comment (ChrisAubuchon) [inline roles/certificates/tasks/main.yml]: Probably should have `mode: 0644`. Currently deployed as `0666`. 
```
-rw-rw-rw-. 1 root root 1728 Jun  7 20:51 cert
-rw-r--r--. 1 root root 1090 Jun  7 20:51 csr.pem
-rw-rw-rw-. 1 root root 1675 Jun  7 20:51 key
```

File: roles/certificates/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    {% if locality != "" %} --locality "{{ locality }}" {% endif %}
+    {% if organization != "" %} --organization "{{ organization }}" {% endif %}
+    {% if unit != "" %} --unit "{{ unit }}" {% endif %}
+    {% if email != "" %} --email "{{ email }}" {% endif %}
+    {% if dns != "" %} --dns "{{ dns | unique | join(' ') }}" {% endif %}
+    {% if ip != "" %} --ip "{{ ip | unique | join(' ') }}" {% endif %}
+    --location "{{ location }}"
+    --destination "{{ destination }}"
+  args:
+    creates: "{{ destination }}/{{ prefix }}{% if prefix != '' %}.{% endif %}key"
```
Comment (ChrisAubuchon) [inline roles/certificates/tasks/main.yml]: The files generated by `generate-certificate` are `0666`. Should look in to adding a command line option to set the permissions and default to `0644`

Comment (siddharthist) [inline roles/certificates/tasks/main.yml]: done in latest commit!

**Comment by distributorofpain** [comment]
It seems on centos, the cacert.pem file also needs to be in the /etc/pki/ca-trust/source directory before the update-ca-trust is run.  I was having trouble with my edge nodes not recognizing the certs created on the hosts when i included the cacert.pem in the certificate.  After copying the cacert.pem file from the anchors directory, up one level, to the source directory, running the update-ca-trust command and restarting traefik, the certificate authority was then recognized as valid.  This allows me to use end to end SSL with the new certificate method being used here.

So for docker apps(or any app in marathon), we grab the host cert and cacert files from the hosts /etc/pki/mantl and /etc/pki/CA/ directories, combine them into one ( cert file being first ), then use the combined certificate file inside the application.  this then allows traefik to successfully connect end to end.

So to be clear, i am asking if this command can be added:
```
cp /etc/pki/ca-trust/source/anchors/cacert.pem /etc/pki/ca-trust/source/cacert.pem
```

if not already run before traefik restart, then this as well, before traefik restart
```
update-ca-trust
```

**Comment by siddharthist** [comment]
@distributorofpain Done: a45eb4f

**Comment by distributorofpain** [comment]
awesome, thank you!

File: roles/certificates/tasks/main.yml (hunk 4 above)
Comment (ChrisAubuchon) [inline roles/certificates/tasks/main.yml]: Why have `{{ prefix }}` if all of the other locations are hardcoded?

Comment (siddharthist) [inline roles/certificates/tasks/main.yml]: The role could be reused in e.g. an addon, where you might want to change `prefix`.

**Comment by ryane** [comment]
I am getting these errors:

```
TASK: [certificates | set permissions on certificate files] *******************
changed: [resching-gce-control-03] => (item={'path': '/etc/pki/mantl/cert', 'mode': 420})
changed: [resching-gce-control-01] => (item={'path': '/etc/pki/mantl/cert', 'mode': 420})
changed: [resching-gce-control-02] => (item={'path': '/etc/pki/mantl/cert', 'mode': 420})
changed: [resching-gce-edge-01] => (item={'path': '/etc/pki/mantl/cert', 'mode': 420})
changed: [resching-gce-infra-worker-02] => (item={'path': '/etc/pki/mantl/cert', 'mode': 420})
changed: [resching-gce-infra-worker-01] => (item={'path': '/etc/pki/mantl/cert', 'mode': 420})
changed: [resching-gce-worker-02] => (item={'path': '/etc/pki/mantl/cert', 'mode': 420})
changed: [resching-gce-worker-01] => (item={'path': '/etc/pki/mantl/cert', 'mode': 420})
changed: [resching-gce-kubeworker-01] => (item={'path': '/etc/pki/mantl/cert', 'mode': 420})
changed: [resching-gce-kubeworker-02] => (item={'path': '/etc/pki/mantl/cert', 'mode': 420})
failed: [resching-gce-control-03] => (item={'path': '/etc/pki/mantl/key', 'mode': 384}) => {"details": "bad symbolic permission for mode: 384", "failed": true, "gid": 0, "group": "root", "item": {"mode": 384, "path": "/etc/pki/mantl/key"}, "mode": "0666", "owner": "root", "path": "/etc/pki/mantl/key", "secontext": "unconfined_u:object_r:cert_t:s0", "size": 1679, "state": "file", "uid": 0}
msg: mode must be in octal or symbolic form
failed: [resching-gce-control-01] => (item={'path': '/etc/pki/mantl/key', 'mode': 384}) => {"details": "bad symbolic permission for mode: 384", "failed": true, "gid": 0, "group": "root", "item": {"mode": 384, "path": "/etc/pki/mantl/key"}, "mode": "0666", "owner": "root", "path": "/etc/pki/mantl/key", "secontext": "unconfined_u:object_r:cert_t:s0", "size": 1679, "state": "file", "uid": 0}
msg: mode must be in octal or symbolic form
failed: [resching-gce-edge-01] => (item={'path': '/etc/pki/mantl/key', 'mode': 384}) => {"details": "bad symbolic permission for mode: 384", "failed": true, "gid": 0, "group": "root", "item": {"mode": 384, "path": "/etc/pki/mantl/key"}, "mode": "0666", "owner": "root", "path": "/etc/pki/mantl/key", "secontext": "unconfined_u:object_r:cert_t:s0", "size": 1675, "state": "file", "uid": 0}
msg: mode must be in octal or symbolic form
failed: [resching-gce-control-02] => (item={'path': '/etc/pki/mantl/key', 'mode': 384}) => {"details": "bad symbolic permission for mode: 384", "failed": true, "gid": 0, "group": "root", "item": {"mode": 384, "path": "/etc/pki/mantl/key"}, "mode": "0666", "owner": "root", "path": "/etc/pki/mantl/key", "secontext": "unconfined_u:object_r:cert_t:s0", "size": 1675, "state": "file", "uid": 0}
msg: mode must be in octal or symbolic form
failed: [resching-gce-kubeworker-01] => (item={'path': '/etc/pki/mantl/key', 'mode': 384}) => {"details": "bad symbolic permission for mode: 384", "failed": true, "gid": 0, "group": "root", "item": {"mode": 384, "path": "/etc/pki/mantl/key"}, "mode": "0666", "owner": "root", "path": "/etc/pki/mantl/key", "secontext": "unconfined_u:object_r:cert_t:s0", "size": 1675, "state": "file", "uid": 0}
msg: mode must be in octal or symbolic form
failed: [resching-gce-worker-02] => (item={'path': '/etc/pki/mantl/key', 'mode': 384}) => {"details": "bad symbolic permission for mode: 384", "failed": true, "gid": 0, "group": "root", "item": {"mode": 384, "path": "/etc/pki/mantl/key"}, "mode": "0666", "owner": "root", "path": "/etc/pki/mantl/key", "secontext": "unconfined_u:object_r:cert_t:s0", "size": 1675, "state": "file", "uid": 0}
msg: mode must be in octal or symbolic form
failed: [resching-gce-worker-01] => (item={'path': '/etc/pki/mantl/key', 'mode': 384}) => {"details": "bad symbolic permission for mode: 384", "failed": true, "gid": 0, "group": "root", "item": {"mode": 384, "path": "/etc/pki/mantl/key"}, "mode": "0666", "owner": "root", "path": "/etc/pki/mantl/key", "secontext": "unconfined_u:object_r:cert_t:s0", "size": 1679, "state": "file", "uid": 0}
msg: mode must be in octal or symbolic form
failed: [resching-gce-infra-worker-01] => (item={'path': '/etc/pki/mantl/key', 'mode': 384}) => {"details": "bad symbolic permission for mode: 384", "failed": true, "gid": 0, "group": "root", "item": {"mode": 384, "path": "/etc/pki/mantl/key"}, "mode": "0666", "owner": "root", "path": "/etc/pki/mantl/key", "secontext": "unconfined_u:object_r:cert_t:s0", "size": 1679, "state": "file", "uid": 0}
msg: mode must be in octal or symbolic form
failed: [resching-gce-infra-worker-02] => (item={'path': '/etc/pki/mantl/key', 'mode': 384}) => {"details": "bad symbolic permission for mode: 384", "failed": true, "gid": 0, "group": "root", "item": {"mode": 384, "path": "/etc/pki/mantl/key"}, "mode": "0666", "owner": "root", "path": "/etc/pki/mantl/key", "secontext": "unconfined_u:object_r:cert_t:s0", "size": 1679, "state": "file", "uid": 0}
msg: mode must be in octal or symbolic form
failed: [resching-gce-kubeworker-02] => (item={'path': '/etc/pki/mantl/key', 'mode': 384}) => {"details": "bad symbolic permission for mode: 384", "failed": true, "gid": 0, "group": "root", "item": {"mode": 384, "path": "/etc/pki/mantl/key"}, "mode": "0666", "owner": "root", "path": "/etc/pki/mantl/key", "secontext": "unconfined_u:object_r:cert_t:s0", "size": 1675, "state": "file", "uid": 0}
msg: mode must be in octal or symbolic form

FATAL: all hosts have already failed -- aborting
```

I am on ansible 1.9.4

**Comment by ryane** [comment]
ci builds failing with same errors

**Comment by siddharthist** [comment]
This isn't ready anymore, I'll post here when it is.

**Comment by siddharthist** [comment]
Okay, I've manually tested on AWS with the latest commit, with and without Docker TCP (with and without TLS) and it all worked for me. This is once more ready for review!

**Comment by siddharthist** [comment]
One of the CI builds got hit by #1372 and another by #1492, but AWS passed!

**Comment by siddharthist** [comment]
This has passed another three CI builds on the latest master, been tested manually on latest master, and been tested on a 1.1 upgrade. Merging.

**Comment by distributorofpain** [comment]
YAY!

**Pull request merged**